src/compiler/transformers/es6.ts

```diff
@@ -52,20 +184,37 @@ namespace ts {
         function transformSourceFile(node: SourceFile) {
             currentSourceFile = node;
             enclosingBlockScopeContainer = node;
+            currentNode = node;
             return visitEachChild(node, visitor, context);
         }
 
         function visitor(node: Node): VisitResult<Node> {
+            return saveStateAndInvoke(node, dispatcher);
+        }
+
+        function dispatcher(node: Node): VisitResult<Node> {
+            return convertedLoopState
+                ? visitorForConvertedLoopWorker(node)
+                : visitorWorker(node);
+        }
+
+        function saveStateAndInvoke<T>(node: Node, f: (node: Node) => T): T {
             const savedContainingNonArrowFunction = containingNonArrowFunction;
             const savedCurrentParent = currentParent;
             const savedCurrentNode = currentNode;
             const savedEnclosingBlockScopeContainer = enclosingBlockScopeContainer;
             const savedEnclosingBlockScopeContainerParent = enclosingBlockScopeContainerParent;
 
-            onBeforeVisitNode(node);
+            const savedConvertedLoopState = convertedLoopState;
+            if (nodeStartsNewLexicalEnvironment(node)) {
+                // don't treat content of nodes that start new lexical environment or class-like nodes as part of converted loop copy 

```

Comment:
comment is fixed. Re initializers, I think this code should be equivalent to 
```ts
"use strict"
let ar = [];
for (let i = 0; i < 2; i++) {
  ar.push(class { constructor() { this.x=i }; });
}
console.log(new ar[0]().x); // should be 0
console.log(new ar[1]().x); // should be 1
```

old emitter does not support this either so I'll file it as a separate issue that should be fixed